Adding alltoall_single collective to collective quantization API

marjanfariborz opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63154

The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Commit:
Adding alltoall_single collective to collective quantization API

The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

wanchaol commented on test/distributed/algorithms/quantization/quantization_test.py:
```diff
@@ -140,3 +170,29 @@ def _test_all_to_all(
+                expected_tensor = expected_tensor.cuda(rank_to_GPU[rank][0])
+                out_tensor = out_tensor.cuda(rank_to_GPU[rank][0])
+            if(qtype is not None):
+                quantize_alltoall_single = quant.auto_quantize(dist.all_to_all_single, qtype, quant_loss=None)
```
qtype must be none for our tests cases right? I think we should make `qtype` not an optional argument as we just want to test the quantization part, if `qtype=None`, the tests should already been covered by `distributed_test.py`

wanchaol commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -114,6 +114,8 @@ def auto_quantize(func, qtype, quant_loss=None):
     """
```
I think you should update the documentation as well.

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -116,6 +116,8 @@ def auto_quantize(func, qtype, quant_loss=None):
     """
     @functools.wraps(func)
     def wrapper(*args, **kwargs):
+        out_splits = kwargs.get('out_splits', None)
```
what's these two variables used for? shouldn't we place these to the section of `all_to_all_single` as it's only specific to `all_to_all_single`?

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -130,3 +162,24 @@ def _test_all_to_all(
+            if cuda:
+                in_tensor = in_tensor.cuda(rank_to_GPU[rank][0])
+                expected_tensor = expected_tensor.cuda(rank_to_GPU[rank][0])
+                out_tensor = out_tensor.cuda(rank_to_GPU[rank][0])
```
nit: save `rank_to_GPU[rank][0]` in a temporary?

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -125,6 +125,16 @@ def wrapper(*args, **kwargs):
             for i, t in enumerate(_dequantize_tensor_list(out_tensors, qtype, quant_loss=quant_loss)):
                 tensors[i] = t
 
+        elif (func == dist.all_to_all_single):
```
nit: to avoid another if/else condition here when additional functions are supported, consider turning this into a dispatch table that maps collectives to their quantized versions:

```
dispatch = {
dist.all_to_all_single: quantized_a2a_single_handler,
dist.all_gather: quantized_all_gather_handler,
....
}
```

rohan-varma reviewed (commented):
Overall looks good to me! Please address @wanchaol's comment and then it should be good to go. Thank you!

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]

Commit:
Update on "Adding alltoall_single collective to collective quantization API"


The collective quantization API now supports alltoall, alltoall_single, and allscatter. The test is also included.

Differential Revision: [D30255251](https://our.internmc.facebook.com/intern/diff/D30255251/)

[ghstack-poisoned]
